src/EditorFeatures/Core/CommandHandlers/GoToNextAndPreviousMethodCommandHandler.cs

```diff
@@ -0,0 +1,188 @@
+            if (!caretPoint.HasValue)
+            {
+                nextHandler();
+                return;
+            }
+
+            int? targetPosition = null;
+            var waitResult = _waitIndicator.Wait(EditorFeaturesResources.Navigating, allowCancel: true, action: waitContext =>
+            {
+                targetPosition = GetTargetPosition(document, caretPoint.Value.Position, next, waitContext.CancellationToken);
+            });
+
+            if (waitResult == WaitIndicatorResult.Canceled || targetPosition == null)
+            {
+                return;
+            }
+
+            textView.TryMoveCaretToAndEnsureVisible(new SnapshotPoint(subjectBuffer.CurrentSnapshot, targetPosition.Value), _outliningManagerService);
+        }
+
+        /// <summary>
+        /// Internal for testing purposes.
+        /// </summary>
+        internal static int? GetTargetPosition(Document document, int caretPosition, bool next, CancellationToken cancellationToken)
+        {
+            var syntaxFactsService = document.GetLanguageService<ISyntaxFactsService>();
+            if (syntaxFactsService == null)
+            {
+                return null;
+            }
+
+            var root = document.GetSyntaxRootAsync(cancellationToken).WaitAndGetResult(cancellationToken);
+            var members = syntaxFactsService.GetMethodLevelMembers(root);
+
+            if (members.Count == 0)
+            {
+                return null;
+            }
+
+            var currentMember = GetMember(syntaxFactsService, caretPosition, root);
+            int indexOfCurrentMember;
+            if (currentMember == null)
+            {
+                // We're not contained in any member currently.  Find the first member to start after our
+                // position and use that.
+                for (indexOfCurrentMember = 0; indexOfCurrentMember < members.Count; indexOfCurrentMember++)
+                {
+                    if (members[indexOfCurrentMember].FullSpan.Start > caretPosition)
+                    {
+                        break;
+                    }
+                }
+
+                if (indexOfCurrentMember == members.Count)
+                {
+                    // After the last member, count ourselves as part of that last member, so that we wrap
+                    // to the first, or back to the beginning of last.
+                    indexOfCurrentMember--;
+                }
+
+                currentMember = members[indexOfCurrentMember];
+            }
+            else
+            {
+                indexOfCurrentMember = members.IndexOf(currentMember);
+                if (indexOfCurrentMember < 0)
+                {
+                    Debug.Fail("Couldn't find current member in members?");
+                    return null;
+                }
+            }
+
+            if (next)
+            {
+                // If we're in leading trivia, we just want to go to the start of this member.
```

Comment:
@pilchie Oh! I relied too much on this comment and not enough on the code. This handles being in leading trivia or just generally being before the member.